inc/class-job.php

```diff
@@ -216,6 +229,142 @@ public static function get_by_site( $site, $include_completed = false, $include_
 		return static::to_instances( $results );
 	}
 
+	/**
+	 * Query jobs database.
+	 *
+	 * Returns an array of Job instances for the current site based
+	 * on the paramaters.
+	 *
+	 * @todo: allow searching within time window for duplicate events.
+	 *
+	 * @param array|\stdClass $args {
+	 *     @param string          $hook      Jobs hook to return. Optional.
+	 *     @param int|string|null $timestamp Timestamp to search for. Optional.
+	 *                                       String shortcuts `future`: > NOW(); `past`: <= NOW()
+	 *     @param array           $args      Cron job arguments.
+	 *     @param int|object      $site      Site to query. Default current site.
+	 *     @param array           $statuses  Job statuses to query. Default to waiting and running.
+	 *     @param int             $limit     Max number of jobs to return. Default 1.
+	 *     @param string          $order     ASC or DESC. Default ASC.
+	 * }
+	 * @return Job[]|WP_Error Jobs on success, error otherwise.
+	 */
+	public static function get_jobs_by_query( $args = [] ) {
+		global $wpdb;
+		$args = (array) $args;
+		$results = [];
+
+		$defaults = [
+			'timestamp' => null,
+			'hook' => null,
+			'args' => [],
+			'site' => get_current_blog_id(),
+			'statuses' => [ 'waiting', 'running' ],
+			'limit' => 1,
+			'order' => 'ASC',
+			'__raw' => false,
+		];
+
+		$args = wp_parse_args( $args, $defaults );
+
+		// Allow passing a site object in
+		if ( is_object( $args['site'] ) && isset( $args['site']->blog_id ) ) {
+			$args['site'] = $args['site']->blog_id;
+		}
+
+		if ( ! is_numeric( $args['site'] ) ) {
+			return new WP_Error( 'cavalcade.job.invalid_site_id' );
+		}
+
+		if ( ! empty( $args['hook'] ) && ! is_string( $args['hook'] ) ) {
+			return new WP_Error( 'cavalcade.job.invalid_hook_name' );
+		}
+
+		if ( ! is_array( $args['args'] ) && ! is_null( $args['args'] ) ) {
+			return new WP_Error( 'cavalcade.job.invalid_event_arguments' );
+		}
+
+		if ( ! is_numeric( $args['limit' ] ) ) {
+			return new WP_Error( 'cavalcade.job.invalid_limit' );
+		}
+
+		if ( $args['timestamp'] === 'future' ) {
+			$timestamp = time();
+			$timestamp_compare = '>';
+		} elseif ( $args['timestamp'] === 'past' ) {
+			$timestamp = time();
+			$timestamp_compare = '<=';
+		}
+
+		$args['limit' ] = absint( $args['limit' ] );
+
+		if ( $args['limit'] > 100 ) {
+			trigger_error( 'Exceeding recommended job search limit of 100' );
```

Comment:
WP Core is pretty insistent about getting the cron array (can't figure out why) so deliberately leaving unlimited  `limit => 0` out of this.

Reply:
I figured out why WP was insisting on this, the `cron` option autoloads and the cavalcade cache is not persistent. Changing one of those settings should remove the full query.